### PR TITLE
[SPARK-38394][BUILD] Upgrade `scala-maven-plugin` to 4.4.0 for Hadoop 3 profile

### PR DESCRIPTION
**Pull request by steveloughran**
### What changes were proposed in this pull request?

This sets scala-maven-plugin.version to 4.4.0 except when the hadoop-2.7
profile is used, because SPARK-36547 shows that only 4.3.0 works there.

### Why are the changes needed?

1. If you try to build against a local snapshot of hadoop trunk with `-Dhadoop.version=3.4.0-SNAPSHOT` the build failes with the error shown in the JIRA.
2. upgrading the scala plugin version fixes this. It is a plugin issue. 
3. the version is made configurable so the hadoop 2.7 profile can switch back to the one which works there.

As to why this only surfaces when compiling hadoop trunk, or why hadoop-2.7 requires the new one -who knows. they both look certificate related, which is interesting. maybe something related to signed JARs?

### Does this PR introduce _any_ user-facing change?

no

### How was this patch tested?

by successfully building spark against a local build of hadoop 3.4.0-SNAPSHOT

### CODE REVIEW

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM. Thank you, @steveloughran .

cc @sunchao , @srowen , @HyukjinKwon , @gengliangwang , @cloud-fan ,  @williamhyun

File: pom.xml
Comment (sunchao) [inline pom.xml:3436]: nit: maybe we should keep the previous warning "Please don't upgrade the version below .."

**Review by sunchao — APPROVED** [review]
LGTM too with one nit

**Comment by srowen** [comment]
Weird, but OK by me

File: pom.xml
Comment (martin-g) [inline pom.xml:169]: Why not 4.5.6 ?
https://search.maven.org/artifact/net.alchim31.maven/scala-maven-plugin/4.5.6/maven-plugin

**Review by HyukjinKwon — APPROVED** [review]

File: pom.xml
Comment (steveloughran) [reply pom.xml:3436]: how about "please be very careful" and links to the different JIRAs, and that the 2.7 one must be left as is?

File: pom.xml
Comment (steveloughran) [reply pom.xml:169]: i reverted to the one which was known to work on spark and hadoop 3, which was enough to unblock me without me having to go near any futher regressions caused by later versions of something which is clearly brittle to changes in *something* (transient dependencies?)

Comment (martin-g) [reply pom.xml:169]: OK. I will try it myself!

File: pom.xml
Comment (dongjoon-hyun) [reply pom.xml:3436]: +1 for that description, @steveloughran .

Comment (sunchao) [reply pom.xml:3436]: Yes SGTM too

**Comment by srowen** [comment]
How far does this need to back-port, 3.1?

**Comment by dongjoon-hyun** [comment]
Since this is required for only developing `Apache Hadoop 3.4`, I'd suggest to have this only at Apache Spark 3.3 which means `master` branch, @srowen .

To @steveloughran and @sunchao , when is Apache Hadoop community going to release Apache Hadoop 3.4.0?

**Comment by srowen** [comment]
Merged to master

**Comment by dongjoon-hyun** [comment]
Oh, we were waiting for addressing this comment.
- https://github.com/apache/spark/pull/35725#discussion_r819028128

**Comment by srowen** [comment]
Ah. Didn't catch that. Do we need to revert or just follow up? I can help with either

**Comment by dongjoon-hyun** [comment]
No problem. A simple follow-up will be enough~

**Comment by steveloughran** [comment]
>, when is Apache Hadoop community going to release Apache Hadoop 3.4.0?

no schedule, i was just trying to do end to end testing on the manifest committer and things wouldn't work. moving (back) to the 4.4.0 compiler fixed things

i do have locally the changes to update the pom comments better, just hadn't merged it as i was dealing with end to end build problems (mvn oom, ...), unrelated to scala. will followup

**Comment by steveloughran** [comment]
see #35753

**Comment by steveloughran** [comment]
> Since this is required for only developing Apache Hadoop 3.4, I'd suggest to have this only at Apache Spark 3.3 which means master branch, @srowen .

oh, and i agree with this. its currently only for people trying to do things across multiple projects.

what's not clear though is "why does it now fail on hadoop trunk?". i suspect some transient dependency change; as all these failures seem related to encryption and certificates. maybe something related to signed jars.

whatever that cause is, there's a risk that any change to hadoop branch-3.3 could suddenly stop the build the same way.

**Pull request closed without merging**
